File: year3/week16/solution/week16.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

// Calculates the majority element of a stream in constant time using the boyer-
// moore majority vote algorithm.
template <typename T>
class StreamingMajorityVote {
 public:
  StreamingMajorityVote() : candidate_{}, votes_(0) {}

  void consume(const T& v) {
    v == candidate_ ? ++votes_ : --votes_;
    if (votes_ <= 0) {
      candidate_ = v;
      votes_ = 0;
    }
  }

  const T& candidate() const { return candidate_; }

 private:
  T candidate_;
  int votes_;
};

template <typename T>
std::unordered_map<T, int> histogram(const std::vector<T>& v) {
  std::unordered_map<T, int> h;
  for (const auto& i : v) {
    ++h[i];
  }
  return h;
}

int main() {
  std::ios_base::sync_with_stdio(false);

  int n_districts, n_votes;
  std::cin >> n_districts >> n_votes;

  // Calculate the majority vote for each district.
  std::vector<StreamingMajorityVote<std::string>> districts(n_districts);
  int district;
  std::string candidate;
  while (std::cin >> candidate >> district) {
    districts[district].consume(candidate);
  }

  // Grab each candidates name.
  std::vector<std::string> district_winners(n_districts);
  std::transform(districts.begin(), districts.end(), district_winners.begin(),
                 [](const auto& d) { return d.candidate(); });

  // Count how many districts each candidate won and output the one who took
  // the most.
  const auto tally = histogram(district_winners);
  const auto winner = std::max_element(
      tally.begin(), tally.end(),
      [](const auto& lhs, const auto& rhs) { return lhs.second < rhs.second; });

  std::cout << winner->first << std::endl;
}

File: year3/week16/solution/week16.py
import sys
from collections import Counter

def parse_vote(line):
    candidate, district = line.strip().split(" ")
    return candidate, int(district)

def consume_vote(district_counts, vote):
    candidate, district = vote
    if district_counts[district] is None:
        district_counts[district] = (candidate, 0)

    current_leader, count = district_counts[district]
    if current_leader == candidate:
        district_counts[district] = (current_leader, count + 1)
    else:
        district_counts[district] = (candidate, 1)

def main():
    n_votes, n_districts = map(int, input().strip().split(" "))
    votes = map(parse_vote, sys.stdin.readlines())
    district_counts = [None for _ in range(n_districts)]
    for v in votes: consume_vote(district_counts, v)
    winners = (x[0] for x in district_counts if x is not None)
    print(Counter(winners).most_common(1)[0][0])

main()
